waiter/src/waiter/metrics_sync.clj

```diff
@@ -68,27 +69,34 @@
                      ", current" [router-ws-key router-id] "request-id is" (:request-id ws-request))
           router-metrics-state)))))
 
+(defn- listen-on-ctrl-chan
+  "Deregister any requests corresponding to request-id on router-ws-key when data is received on ctrl channel."
+  [ctrl router-ws-key router-id request-id encrypt router-metrics-agent]
+  (async/go
+    (when-let [ctrl-data (async/<! ctrl)]
+      (cid/cinfo request-id "deregistering request, data received on control channel is" ctrl-data)
+      (send router-metrics-agent deregister-router-ws router-ws-key router-id request-id encrypt))))
+
 (defn register-router-ws
   "Registers the websocket request with the specified request-id into the agent's state.
    It also attaches a callback to deregister the request when the connection receives data on the `ctrl` channel."
-  [{:keys [router-id] :as router-metrics-state} router-ws-key source-router-id {:keys [ctrl request-id] :as ws-request} encrypt router-metrics-agent]
+  [router-metrics-state router-ws-key router-id {:keys [ctrl request-id] :as ws-request} encrypt router-metrics-agent]
   (with-catch
     router-metrics-state
     (if request-id
       (do
-        (cid/cdebug request-id "registering request from router" source-router-id)
-        (counters/inc! (metrics/waiter-counter "metrics-syncer" (router-ws-key->name router-ws-key) (str source-router-id) "register"))
-        (when ctrl
-          (let [control-mult (async/mult ctrl)
-                request-terminate-chan (async/tap control-mult (au/sliding-buffer-chan 1))]
-            (async/go
-              (when-let [ctrl-data (async/<! request-terminate-chan)]
-                (cid/cinfo request-id "received" ctrl-data "in control channel, deregistering request")
-                (send router-metrics-agent deregister-router-ws router-ws-key source-router-id request-id encrypt)))))
-        (assoc-in router-metrics-state [router-ws-key source-router-id]
-                  (select-keys ws-request [:in :ctrl :out :request-id :tag])))
+        (cid/cinfo request-id "registering" (name router-ws-key) "request" router-id)
+        (counters/inc! (metrics/waiter-counter "metrics-syncer" (router-ws-key->name router-ws-key) router-id "register"))
+        (if ctrl
+          (do
+            (listen-on-ctrl-chan ctrl router-ws-key router-id request-id encrypt router-metrics-agent)
```

Comment:
Why not limit the call to `listen-on-ctrl-chan` to `update-router-metrics-state`, since that will run in all cases?